File: python/anthropic/claude_tools.py
import json
import os
import random
from anthropic import Anthropic
from anthropic.types import ToolUseBlock, ToolResultBlockParam

from dotenv import load_dotenv

load_dotenv()

client = Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))

MODEL = "claude-3-5-sonnet-20241022"

SYS_PROMPT = """
You are a friendly chatbot answering the user's questions.
"""

user_msg = "What's the weather like in Helsinki?"


tools = [
    {
        "name": "getWeather",
        "description": "Get weather information about a city",
        "input_schema": {
            "type": "object",
            "properties": {
                "city": {
                    "type": "string",
                    "description": "Name of the city",
                }
            },
            "required": ["city"],
        },
    }
]


def handle_tool_call(tool_call: ToolUseBlock):
    tool_name = tool_call.name

    print(f"LLM called tool: {tool_name} with args: {tool_call.input}")

    temperature = random.randint(0, 20)
    weather_type = random.choice(["sunny", "cloudy", "raining"])

    if tool_name == "getWeather":
        city = tool_call.input["city"]

        return ToolResultBlockParam(
            type="tool_result",
            tool_use_id=tool_call.id,
            content=f"The weather in {city} today: {temperature}°, {weather_type}",
        )
    else:
        raise ValueError(f"Unknown tool {tool_name}")


def call_llm(messages, tools=None):
    response = client.messages.create(
        model=MODEL,
        max_tokens=1024,
        messages=messages,
        system=SYS_PROMPT,
        tools=tools,
    )
    return response


def stream_llm(messages, tools=None):
    with client.messages.stream(
        model=MODEL,
        max_tokens=1024,
        messages=messages,
        system=SYS_PROMPT,
        tools=tools,
    ) as stream:
        answer = ""
        for text in stream.text_stream:
            answer += text or ""
            print(text, end="", flush=True)

        return answer


messages = [{"role": "user", "content": user_msg}]

response = call_llm(messages, tools=tools)

if response.stop_reason == "end_turn":
    print(response.content[0].text)
elif response.stop_reason == "tool_use":
    messages.append({"role": "assistant", "content": response.content})

    for block in response.content:
        if block.type == "text":
            print(block.text)
        elif block.type == "tool_use":
            tool_result = handle_tool_call(block)
            messages.append({"role": "user", "content": [tool_result]})

    stream_llm(messages=messages, tools=tools)

File: node/src/anthropic/claude_tools.ts
import Anthropic from "@anthropic-ai/sdk";
import { Tool } from "@anthropic-ai/sdk/resources/messages.mjs";
import dotenv from "dotenv";
import path from "path";

dotenv.config({ path: path.resolve(__dirname, "../../.env") });

const anthropic = new Anthropic();

const MODEL = "claude-3-5-sonnet-20241022";

const SYS_PROMPT = `
You are a friendly chatbot answering the user's questions.
`;

const tools: Tool[] = [
  {
    name: "getWeather",
    description: "Get weather information about a city",
    input_schema: {
      type: "object",
      properties: {
        city: {
          type: "string",
          description: "Name of the city",
        },
      },
      required: ["city"],
    },
  },
];

const user_msg = "What's the weather like in Helsinki?";

const messages: Anthropic.Messages.MessageParam[] = [
  { role: "user", content: user_msg },
];

const handleToolCall = (toolCall: Anthropic.Messages.ToolUseBlock) => {
  const toolArgs = toolCall.input as { city: string };

  if (toolCall.name === "getWeather") {
    const temperature = Math.floor(Math.random() * 21);
    const weather_types = ["sunny", "cloudy", "raining"];
    const weather_type =
      weather_types[Math.floor(Math.random() * weather_types.length)];

    const city = toolArgs.city;

    const toolResult: Anthropic.Messages.ToolResultBlockParam = {
      type: "tool_result",
      tool_use_id: toolCall.id,
      content: `The weather in ${city} today: ${temperature}°, ${weather_type}`,
    };
    return toolResult;
  } else {
    throw new Error(`Unknown tool ${toolCall.name}`);
  }
};

const askWeather = async () => {
  const response = await anthropic.messages.create({
    system: SYS_PROMPT,
    messages: messages,
    model: MODEL,
    max_tokens: 1024,
    tools: tools,
  });

  if (response.stop_reason == "end_turn") {
    if (response.content[0].type == "text") {
      console.log(response.content[0].text);
    }
  } else if (response.stop_reason == "tool_use") {
    messages.push({ role: "assistant", content: response.content });

    const blocks: Anthropic.Messages.ContentBlock[] = response.content;

    for (const block of blocks) {
      if (typeof block === "string") {
        console.log(block);
      } else if (block.type === "text") {
        console.log(block.text);
      } else if (block.type === "tool_use") {
        // Handle tool use block
        console.log(`Tool called: ${block.name}`);
        console.log(`Tool input: ${JSON.stringify(block.input)}`);

        const tool_result = handleToolCall(block);
        messages.push({ role: "user", content: [tool_result] });
      }
    }

    const stream = anthropic.messages.stream({
      system: SYS_PROMPT,
      messages,
      model: MODEL,
      max_tokens: 1024,
      tools: tools,
    });

    for await (const block of stream) {
      if (block.type === "content_block_delta") {
        process.stdout.write(
          (block.delta as Anthropic.Messages.TextDelta).text || ""
        );
      }
    }
  }
};

askWeather();
